pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,444 @@
+	MaxChunksGetRangeRequests int `yaml:"max_chunks_get_range_requests"`
+
+	// TTLs for various cache items.
+	ChunkObjectSizeTTL time.Duration `yaml:"chunk_object_size_ttl"`
+	ChunkBlockTTL      time.Duration `yaml:"chunk_block_ttl"`
+}
+
+func DefaultCachingBucketConfig() CachingBucketConfig {
+	return CachingBucketConfig{
+		ChunkBlockSize:            16000, // Equal to max chunk size.
+		ChunkObjectSizeTTL:        24 * time.Hour,
+		ChunkBlockTTL:             24 * time.Hour,
+		MaxChunksGetRangeRequests: 3,
+	}
+}
+
+// Bucket implementation that provides some caching features, using knowledge about how Thanos accesses data.
+type CachingBucket struct {
+	bucket objstore.Bucket
+	cache  cache.Cache
+
+	config CachingBucketConfig
+
+	logger log.Logger
+
+	requestedChunkBytes prometheus.Counter
+	fetchedChunkBytes   *prometheus.CounterVec
+	refetchedChunkBytes *prometheus.CounterVec
+
+	objectSizeRequests prometheus.Counter
+	objectSizeHits     prometheus.Counter
+}
+
+func NewCachingBucket(b objstore.Bucket, c cache.Cache, chunks CachingBucketConfig, logger log.Logger, reg prometheus.Registerer) (*CachingBucket, error) {
+	if b == nil {
+		return nil, errors.New("bucket is nil")
+	}
+	if c == nil {
+		return nil, errors.New("cache is nil")
+	}
+
+	cb := &CachingBucket{
+		config: chunks,
+		bucket: b,
+		cache:  c,
+		logger: logger,
+
+		requestedChunkBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_requested_chunk_bytes_total",
+			Help: "Total number of requested bytes for chunk data.",
+		}),
+		fetchedChunkBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_fetched_chunk_bytes_total",
+			Help: "Total number of fetched chunk bytes. Data from bucket is then stored to cache.",
+		}, []string{"origin"}),
+		refetchedChunkBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_refetched_chunk_bytes_total",
+			Help: "Total number of chunk bytes re-fetched from storage, despite being in cache already.",
+		}, []string{"origin"}),
+		objectSizeRequests: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_objectsize_requests_total",
+			Help: "Number of object size requests for objects.",
+		}),
+		objectSizeHits: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_objectsize_hits_total",
+			Help: "Number of object size hits for objects.",
+		}),
+	}
+
+	cb.fetchedChunkBytes.WithLabelValues(originBucket)
+	cb.fetchedChunkBytes.WithLabelValues(originCache)
+	cb.refetchedChunkBytes.WithLabelValues(originCache)
+
+	return cb, nil
+}
+
+func (cb *CachingBucket) Close() error {
+	return cb.bucket.Close()
+}
+
+func (cb *CachingBucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	return cb.bucket.Upload(ctx, name, r)
```

Comment:
What do you think about embedding objstore.Bucket so we can totally avoid writting those boilerplate methods?

Reply:
It is extra boilerplate, but I think it's also cleaner design-wise. Plus I'm going to add logic to other methods in next PR (for metadata caching).

Reply:
This is fine, you can overwrite the method always. I think it really depends what is the consistent approach. If we have boilerplate like this and only one method is changes we always embed, but it's up to you here, not a blocker.

Reply:
But hm I would really just do embedding. The key here is that is will show us CLEARLY what caching is adding on top of normal obj store ):

Reply:
OK

Reply:
embedding will expose underlying bucket in a public field, instead of private (not a big deal, just saying)